usage/src/org/apache/cloudstack/quota/QuotaAlertManagerImpl.java

```diff
@@ -0,0 +1,444 @@
+            final List<String> emailRecipients = new ArrayList<String>();
+            for (UserVO user : usersInAccount) {
+                userNames += String.format("%s <%s>,", user.getUsername(), user.getEmail());
+                emailRecipients.add(user.getEmail());
+            }
+            if (userNames.endsWith(",")) {
+                userNames = userNames.substring(0, userNames.length() - 1);
+            }
+
+            final Map<String, String> optionMap = new HashMap<String, String>();
+            optionMap.put("accountName", account.getAccountName());
+            optionMap.put("accountID", account.getUuid());
+            optionMap.put("accountUsers", userNames);
+            optionMap.put("domainName", accountDomain.getName());
+            optionMap.put("domainID", accountDomain.getUuid());
+            optionMap.put("quotaBalance", QuotaConfig.QuotaCurrencySymbol.value() + " " + balance.toString());
+            if (emailType == QuotaEmailTemplateTypes.QUOTA_STATEMENT) {
+                optionMap.put("quotaUsage", QuotaConfig.QuotaCurrencySymbol.value() + " " + usage.toString());
+            }
+
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug("accountName" + account.getAccountName() + "accountID" + account.getUuid() + "accountUsers" + userNames + "domainName" + accountDomain.getName() + "domainID"
+                        + accountDomain.getUuid());
+            }
+
+            final StrSubstitutor templateEngine = new StrSubstitutor(optionMap);
+            final String subject = templateEngine.replace(emailTemplate.getTemplateSubject());
+            final String body = templateEngine.replace(emailTemplate.getTemplateBody());
+            try {
+                _emailQuotaAlert.sendQuotaAlert(emailRecipients, subject, body);
+                emailToBeSent.sentSuccessfully(_quotaAcc);
+            } catch (Exception e) {
+                s_logger.error(String.format("Unable to send quota alert email (subject=%s; body=%s) to account %s (%s) recipients (%s) due to error (%s)", subject, body, account.getAccountName(),
+                        account.getUuid(), emailRecipients, e));
+            }
+        } else {
+            s_logger.error(String.format("No quota email template found for type %s, cannot send quota alert email to account %s(%s)", emailType, account.getAccountName(), account.getUuid()));
+        }
+    }
+
+    public static long getDifferenceDays(Date d1, Date d2) {
+        long diff = d2.getTime() - d1.getTime();
+        return TimeUnit.DAYS.convert(diff, TimeUnit.MILLISECONDS);
+    }
+
+    protected boolean lockAccount(long accountId) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.CLOUD_DB).close();
+        boolean success = false;
+        Account account = _accountDao.findById(accountId);
+        if (account != null) {
+            if (account.getState().equals(State.locked)) {
```

Comment:
Enumeration values are guarenteed singletons.  Therefore, ``==`` is preferred over ``equals``.